Streamline dependency generation

levitte opened the pull request:
It seems that only gcc -MMD produces dependency files that are "sane"
for our needs.  For all other methods, some post processing is needed:

- 'makedepend' (Unix) insists that object files are located in the
  same spot as the source file.
- 'cl /Zs /showIncludes' (Visual C) has "Note: including file: " where
  we'd like to see the object.
- 'CC/DECC' (VMS) insists that the object file is located in the
  current directory, i.e. it strips away all directory information.

So far, we've managed this (except for the VMS case) with individual
uncommented perl command lines directly in the build file template.
We're now collecting these diverse hacks into one perl script that
takes an argument to tell what kind of input to expect and that
massages whatever it gets on STDIN and outputs the result on STDOUT.

levitte committed:
Streamline dependency generation

It seems that only gcc -MMD produces dependency files that are "sane"
for our needs.  For all other methods, some post processing is needed:

- 'makedepend' (Unix) insists that object files are located in the
  same spot as the source file.
- 'cl /Zs /showIncludes' (Visual C) has "Note: including file: " where
  we'd like to see the object.
- 'CC/DECC' (VMS) insists that the object file is located in the
  current directory, i.e. it strips away all directory information.

So far, we've managed this (except for the VMS case) with individual
uncommented perl command lines directly in the build file template.
We're now collecting these diverse hacks into one perl script that
takes an argument to tell what kind of input to expect and that
massages whatever it gets on STDIN and outputs the result on STDOUT.

levitte commented:
This appears to work everywhere (including VMS).  Time for review.

levitte commented on Configurations/descrip.mms.tmpl:
```diff
@@ -900,9 +904,8 @@ $obj.OBJ : $deps
-                 RENAME $obj.tmp-D $obj.d )
-        \@ IF F\$SEARCH("$obj.tmp-D") .NES. "" THEN DELETE $obj.tmp-D;*
+        - \$(PERL) $postprocess_makedepend "VMS C" $objdir < $obj.tmp-D > $obj.d
+        - DELETE $obj.tmp-D;*
```
Someone might want to know why I removed the compare-and-rename thing.  The reason is simple, we don't depend directly on the .d files anywhere, so there's no point being careful with the time stamps...

richsalz commented on util/postprocess-makedepend.pl:
```diff
@@ -0,0 +1,75 @@
+        my $object = $ARGV[1];
+        s/^Note: including file: */${object}: /;
+
+    }
```
I think there should be an explicit no-op which just copies things or an error if it's not one of the supported types.

levitte replied:
An error would be the better choice, methinks, 'cause that's would be on us.  Good point, will fix.

levitte committed:
fixup! Streamline dependency generation

levitte committed:
fixup! Streamline dependency generation

dot-asm commented on util/postprocess-makedepend.pl:
```diff
@@ -0,0 +1,81 @@
+
+    } elsif ($producer eq 'Windows C') {
+
+        # For the moment, we only know of one native Windows C compiler, and
```
This is not true. As far as actual *knowing* goes you can't deny that you, Richard, know of Intel compiler for Windows :-) :-) :-) What I'm saying is wouldn't it be more appropriate to say that we *support* only one compiler. And it would be *more* than appropriate to not refer to it as "Windows C", but by its real name, Visual C, or MSVC, or VC. Yes, one can argue that just like gcc sets the tone in Unix, VC does the same on Windows, but I'd still be against referring to it as "Windows C". Basically it reads as if there is Windows C *language*, and there is not such thing. Yes, "VMS C" has the same problem. In other words these would be explicit reference to tool-chains. This is actually good example of what I talked about last time in context of template neutrality. My line of reasoning was that one should put weight on tool-chain-specific rules, as opposite to producing infinite amount of parameters. That time I called out at manifest embedding, which could be implemented rather as VC-specific post-link rule addendum, than assuming that every tool-chain has equivalent replacement. Because it's not actually given that every tool-chain even has a post-link step. Same can be said about this dependency thing... Well, as you should know, I'm not on good terms with words, so all this might appear as meaningless rambling... But I do have couple of more concrete remarks, coming next. But here I do suggest to make arguments explicitly about tool-chains.

levitte replied:
Actually, "VMS C" is as valid as anything, it's what many people are commonly calling what once was ~~DEC~~ / ~~Compaq~~ / HP / VSI C these days (see the problem here? ;-)  some jokingly talked about DECompHP C, and the VSI came along...).  Basically, if you say "VMS C" to VMS folks, they know *exactly* what you're talking about.

But ok, I can go for VC, considering we have some hard-coded VC things.

Side note re this:
> My line of reasoning was that one should put weight on tool-chain-specific rules, as opposite to producing infinite amount of parameters.

I agree with you, and I have some thoughts on how to do that...  but what I have in mind is a *major* change of the config attributes and the build file templates...  To be talked about separately.

dot-asm commented on util/postprocess-makedepend.pl:
```diff
@@ -0,0 +1,81 @@
+        # with an object file specification, given to us as $ARGV[1].
+
+        my $object = $ARGV[1];
+        s/^Note: including file: */${object}: /;
```
If I look at generated file I spot things like `..\openssl\crypto\some\thing.c(100): warning C4267 ...`. In other words it looks like one should omit all non-Note lines.

Secondly it lists all the system header files, and I don't think this is actually intentional. Of course problem is how to distinguish system header files from OpenSSL ones. Trouble there is that it produces absolute paths even for files that you include with relative reference. I mean you say -I..\openssl, but /showIncludes produces absolute path. And so you have mixture of `c:\program files\windows kits\10\some\thing` and `c:\users\user\some\where\openssl`. Is it possible to use SRCDIR and BLDIR somehow?

dot-asm replied:
However! I couldn't help noticing that all those .d files ... are not used for anything, as depend: rule in the makefile is empty.

levitte replied:
> If I look at generated file I spot things like `..\openssl\crypto\some\thing.c(100): warning C4267 ...`. In other words it looks like one should omit all non-Note lines.

Ah, of course, that's why the previous hack had this collection thing going on.  I totally forgot!  Will fix.

> Secondly it lists all the system header files, and I don't think this is actually intentional.

I know, and I left it like that for exactly the reasons you mention, that it's difficult to distinguish them from our own.  I have the exact same kind of problem on VMS, although, the system header files should be possible to detect, 'cause they all contain a `$` and are actually text libraries (all header files enclosed), extension `.TLB`.

> Is it possible to use SRCDIR and BLDIR somehow?

It would if they were absolute, but they might not be...  they are definitely relative when on the same device.

levitte replied:
> However! I couldn't help noticing that all those .d files ... are not used for anything, as depend: rule in the makefile is empty.

Ah, one thing I forgot about!  Well, cmd isn't as easy to script as bash, or even DCL, so I haven't quite figured out how to do this thing on Windows.  Do you think you can lend a hand?

dot-asm replied:
> > Is it possible to use SRCDIR and BLDIR somehow?

> they are relative

And there is no Perl method to convert them to absolute?

levitte replied:
There is, and we can certainly construct a couple of variables that are guaranteed to hold the absolute paths.
Not by the computer now, will do tomorrow.

dot-asm replied:
As for adding dependency lists to makefile. So far perl served as substitute in such situations. There is File::Find that should be usable as `find $(DEPS) -newer Makefile`. One probably could use same script on all systems...

levitte committed:
fixup! Streamline dependency generation

dot-asm commented on util/postprocess-makedepend.pl:
```diff
@@ -0,0 +1,95 @@
+            die "Producer unrecognised: $producer\n";
+        } else {
+            die "Producer not given\n";
+        }
```
This is just ugly. Checking for command line argument validity in the loop. I understand that the condition is met only once, but that's not how you'd expect program look like... You kind of validate input, then loop. Actually one can say the same about overall structure, in sense that you don't expect *that* kind of condition in the loop... Well, I want to say this is a side note, but just can't collect myself :-)

levitte replied:
You're right. I'll look into refactoring tomorrow...

levitte committed:
fixup! Streamline dependency generation

Refactor util/postprocess-makedepend to be more efficient

levitte commented:
`util/postprocess-makedepend.pl` refactored

levitte committed:
fixup! Streamline dependency generation

dot-asm commented on util/postprocess-makedepend.pl:
```diff
@@ -0,0 +1,116 @@
+            # the Note:
+
+            if (/^Note: including file: */) {
+                $line = "${object}: ".$';
```
`$'` needs to be quoted. Spaces in `c:\program files\windows kits\10\...` don't play out well.

levitte replied:
Ah, right

levitte committed:
fixup! Streamline dependency generation

util/postprocess-makedepend.pl: For VC, make sure to quote dependencies

dot-asm commented:
Could you rebase this?

levitte commented:
Done...  and off for today

dot-asm commented:
Where do we stand on omitting system headers? Things is that with excessive dependency lists it's getting unbearably slow.

t-j-h commented:
I see no point in including system headers in dependency lists myself ...

levitte commented:
Ah, thanks for the reminder

levitte committed:
fixup! Streamline dependency generation

Remake the procedures to return undef directly to discard a line

levitte committed:
util/postprocess-makedepend.pl: For VC, don't include system headers

All dependencies that VC gives us are absolute paths, so we need to
check if some of them are within our source or build tree.  We do that
by comparing the start of each dependency with the absolute versions
of our source and build directories.

levitte commented:
Okie, added a couple of commits, I hope they do things right (haven't tested yet)

dot-asm commented on util/postprocess-makedepend.pl:
```diff
@@ -0,0 +1,134 @@
+                my $substr2 = substr($tail, 0, length $abs_blddir);
+
+                if (lc $abs_srcdir eq lc $substr1
+                        || lc $abs_blddir eq lc $substr2) {
```
Wouldn't `$tail !~ m|^$abs_scrdir|i && $tail !~ m|^$abs_blddir|i` do?

levitte replied:
Can you guarantee that they contain safe characters?  I can't.

dot-asm replied:
`\Q` `\E`

dot-asm replied:
> \Q \E

no, of course not...

dot-asm replied:
> \Q \E

Oh yes, \Q \E works, but original condition should be reversed, i.e. `$tail \~ m|^\Q$abs_scrdir\E|i || $tail =~ m|^\Q$abs_blddir\E|i`.

levitte replied:
Actually, that should work, shouldn't it?  Within what the perl version in question supports re non-ASCII characters and all that...  but still, `\Q` and `\E` existed before 5.10, and were functional as far as I know.  Do you have a different experience?

levitte replied:
Ah, now I see that you figured that out...

levitte replied:
Fixed

dot-asm commented on Configurations/unix-Makefile.tmpl:
```diff
@@ -1032,10 +1032,10 @@ $obj$objext: $deps
+	\$(MAKEDEPEND) -f- -o"|\$\@" -- $incs $cmdflags -- $srcs 2>/dev/null \\
+	    | \$(PERL) \$(SRCDIR)/util/postprocess-makedepend.pl \\
+	               'makedepend' > $obj$depext.tmp
+	if cmp $obj$depext.tmp $obj$depext > /dev/null 2> /dev/null; then \\
```
Why is `@` omitted?

levitte replied:
Because I find it confusing to see an output to `foo/blah.d.tmp` and then not see what happens with it.

dot-asm replied:
Since it, comparison, is not needed in other cases, maybe one can omit it even here? Or what am I missing? [I can't think quite straight yet.]

levitte replied:
You're missing nothing, that's actually a good point...

dot-asm commented on util/postprocess-makedepend.pl:
```diff
@@ -0,0 +1,134 @@
+die "Producer unrecognised: $producer\n" unless defined $procedure;
+
+while (<STDIN>) {
+    $_ = $procedure->($_, @ARGV);
```
How about
```
if ($_ = $procedure->($_, @ARGV)) {
    print or die "$!\n";
}
```
and omit the continue thing. It just stings C eye so bad...

levitte replied:
You gotta switch to the Perl brain when reading Perl ;-)

But sure, I can do that...

levitte replied:
How about this?

``` perl
    $_ = $procedure->($_, @ARGV) and print or print or die "$!\n";
```

dot-asm replied:
Problem with this one is that you have to strain your brain over operator priority. And I think I recall that 'and' and '&&' are different when it comes to priority...

levitte committed:
fixup! Streamline dependency generation

dot-asm commented on util/postprocess-makedepend.pl:
```diff
@@ -0,0 +1,132 @@
+
+while (<STDIN>) {
+    if ($_ = $procedure->($_, @ARGV)) {
+        print or print or die "$!\n";
```
I don't follow. What is it print or print?

levitte replied:
Typoe.  FIxed

levitte committed:
fixup! Streamline dependency generation

levitte committed:
fixup! util/postprocess-makedepend.pl: For VC, don't include system headers

dot-asm commented:
Following kind of belongs in #5606, but problem with Windows is that it seems to update makefile all the time. It might happen that it's problematic or potentially problematic on other OSes, because problem seems to be that list of dependencies to append to makefile comes out in different order all the time, which I *assume* has to do with perl's way to enumerate hashes???

On related side[!] note, as for being sluggish. It's much better now, without system headers, but still slow. One can wonder if nmake would be more happy to see
```
file.obj: one.h \
    two.h
```
rather than
```
file.obj: one.h
file.obj: two.h
```
Though I myself is quite accustomed to configure with no-makedepend, so I don't really care...

t-j-h commented:
Richard should be able to sort the hash and get deterministic output. And combined rules should make it more efficient rather than separate ones - but that is from memory - not from actually testing it.

levitte commented:
#5617 fixes that

levitte commented:
The efficiency of combined rules vs one rule per dependency highly depends on the `make` implementation, doesn't it?

levitte commented:
Collecting them all and outputting the lot at the end should be fairly simple, though...  I'll give it a shot.

levitte committed:
fixup! util/postprocess-makedepend.pl: For VC, don't include system headers

levitte committed:
util/postprocess-makedepend.pl: make an effort to collect dependencies

Instead of just working line by line, we collect all dependencies for
every target and print everything out at the end, with each target
getting a potentially long list of dependencies.

levitte commented:
Done.

levitte committed:
fixup! Streamline dependency generation

dot-asm commented on util/postprocess-makedepend.pl:
```diff
@@ -0,0 +1,160 @@
+        my $buf = $prefix;
+        $prefix = ' ';
+
+        while (@deps && length($buf) + length($deps[0]) <= 77) {
```
That <= 77 sends it to infinite loop if include file name is longer. And it's easily happens with [Windows] absolute pathnames. If there is only one file on line, there shouldn't be *any* limitation on length.

levitte replied:
Ah, good point, will fix.

dot-asm commented on util/postprocess-makedepend.pl:
```diff
@@ -0,0 +1,160 @@
+
+die "Producer unrecognised: $producer\n" unless defined $continuation;
+
+foreach my $target (sort keys %collect) {
```
How many keys can be in %collect? I mean isn't meaning that there will be just one? But even if there can be more than one (on VMS?), wouldn't its value being a list instead of hash be more efficient? That list doesn't have to be sorted, because compiler will always return it in same order...

levitte replied:
> How many keys can be in %collect? I mean isn't meaning that there will be just one?

That's right now, yeah, but I'm trying to make this script generic enough so that if we change our minds on that at some point, it can already handle more than one target.  Boom.

> wouldn't its value being a list instead of hash be more efficient?

Well...  see this for example:

> ```
> : ; makedepend -f- -o'|apps/opt.o' -- -I ../master/ -I ../master/include/ -I include/ -- ../master/apps/opt.c 2>/dev/null
> # DO NOT DELETE
> 
> ../master/apps/opt|apps/opt.o: ../master/apps/apps.h ../master/e_os.h
> ../master/apps/opt|apps/opt.o: /usr/include/limits.h
> ../master/apps/opt|apps/opt.o: /usr/include/bits/libc-header-start.h
> ../master/apps/opt|apps/opt.o: /usr/include/features.h
> ../master/apps/opt|apps/opt.o: /usr/include/stdc-predef.h
> ../master/apps/opt|apps/opt.o: /usr/include/sys/cdefs.h
> ../master/apps/opt|apps/opt.o: /usr/include/bits/wordsize.h
> ...
> ../master/apps/opt|apps/opt.o: ../master/include/openssl/engineerr.h
> ../master/apps/opt|apps/opt.o: ../master/include/openssl/ocsp.h
> ../master/apps/opt|apps/opt.o: ../master/include/openssl/x509v3.h
> ../master/apps/opt|apps/opt.o: ../master/include/openssl/x509v3err.h
> ../master/apps/opt|apps/opt.o: ../master/include/openssl/ocsperr.h
> ...
> ```

If we collect that in a list, we gotta handle repeated targets yada yada yada...  much simpler to use a hash, which isn't *that* inefficient.

levitte replied:
Oh, did you mean that I made a hash of the dependencies as well?  What if any of them is repeated, wouldn't you want to shave that off as well?

dot-asm replied:
> What if any of them is repeated

Fair enough.

levitte committed:
fixup! util/postprocess-makedepend.pl: make an effort to collect dependencies

If a line to be printed contains nothing, add at least one dependency.

dot-asm commented on util/postprocess-makedepend.pl:
```diff
@@ -0,0 +1,160 @@
+            # Remove everything up to the first |
+            $line =~ s/^.*\|//;
+            # Also, remove any dependency that starts with a /, because those
+            # are typically system headers
```
But it doesn't list system headers... So in a sense it's a misleading comment... Maybe it would appropriate to add "in case they are added"...

levitte replied:
Err, say what? `makedepend` does list system headers too, see the example I gave above...

dot-asm replied:
Hmmmm, mine didn't... At least I didn't see it when commented, but then it was late... OK, disregard this for now.

levitte replied:
Did you happen to have added `-Y` when you tried?  I just discovered that flag, and sure enough, then there are no system headers...

dot-asm replied:
Just in case, the question that triggered me to look into it was "what if user uses absolute pathname to Configure". I mean I [for one] am accustomed to `mkdir foo; cd foo; ..\openssl\config; make`, but just might be just me, as I've met people who prefer absolute path names. Maybe it's appropriate to recognize that dependencies are not that meaningful to those who just build and use OpenSSL. They are meaningful only to active OpenSSL developers. And they probably manage, because they are more likely to use gcc. In other words one can as well say that it wouldn't be inappropriate to leave it as it is now.

dot-asm commented on util/postprocess-makedepend.pl:
```diff
@@ -0,0 +1,160 @@
+                # they don't match $abs_srcdir or $abs_blddir
+                $tail = canonpath($tail);
+                if ($tail =~ m|^\Q$abs_srcdir\E|i
+                        || $tail =~ m|^\Q$abs_blddir\E|i) {
```
I'd like to suggest [one last thing]
```
if ($tail =~ s|^\Q$abs_blddir\E|\$(BLDDIR)|i || tail =~ s|^\Q$abs_srcdir\E|$(SRCDIR)|i)
```
Yes, there is ambiguity when you build in source directory, but I'd say it worths it. In case you wonder BLDDIR being moved forward is not a coincidence, but way to make it work when you build in subdirectory of the source directory.

levitte replied:
There is internal knowledged on what files are being generated (i.e. should be in `$(BLDDIR)`) and which ones are not (i.e. should be in `$(SRCDIR)`)...  so it wouldn't be *extremely* hard to get those right.

dot-asm replied:
In such case note that this can just as well be said about makedepend [and presumably VMS]. However! I'd like to suggest to *not* take that path, because it might turn out more expensive than the value it adds. "More expensive" means that it would have to take in information about *all* files or per-file basis. "Per-file" means you execute this script for each file you compile.

dot-asm replied:
I'd say that if you choose to fold/reduce/canonize pathnames in dependency lists, then add-depends.pl would probably be better place.

levitte replied:
This is surely worth looking at for VMS as well...  but not the use of `$(SRCDIR)` and `$(BLDDIR)`, as directory merging isn't as simple as on Unix or Windows...

Regarding the information, this script is already sucking in `configdata.pm`, and `keys %{$unified_info{generated}}` is all the information you need...

levitte replied:
> I'd say that if you choose to fold/reduce/canonize pathnames in dependency lists, then add-depends.pl would probably be better place.

Good point, performance wise.

levitte replied:
I assume that includes what you suggested here, btw.

dot-asm replied:
> I assume that includes what you suggested here, btw.

I don't understand this sentence. Do you mean that you suggest to leave it as it is here and now and do "canonization" in add-depends.pl on all platforms? It would do it, but since I didn't quite get it, could you be more explicit?

dot-asm replied:
As for sucking in configdata.pm and performance. One has to recognize that it's easy to fall victim to illusion that resources are unlimited. But it doesn't change the fact that it's still an illusion. One just adds lines of code without much of a consideration, and then one day you step back and say [what happened?](https://github.com/openssl/openssl/issues/5585). In this context one can actually wonder if it would be appropriate to even split configdata.pm to something minimalistic and say fileinfo.pm. Or somehow arrange that the hash that lists files is not actually instantiated unless you ask for it...

levitte replied:
> I don't understand this sentence. Do you mean that you suggest to leave it as it is here and now and do "canonization" in add-depends.pl on all platforms? It would do it, but since I didn't quite get it, could you be more explicit?

At the beginning of this comment thread, you suggest this change:

```
if ($tail =~ s|^\Q$abs_blddir\E|\$(BLDDIR)|i || tail =~ s|^\Q$abs_srcdir\E|$(SRCDIR)|i)
```

I assume that this would also do better in `add-depends.pl`

levitte replied:
> In this context one can actually wonder if it would be appropriate to even split configdata.pm to something minimalistic and say fileinfo.pm.

It might be a bit late in the game to refactor at that level, at least for 1.1.1

> Or somehow arrange that the hash that lists files is not actually instantiated unless you ask for it...

Soooo....  you're saying that I should qualify the use of `configdata` a bit?  Like this?

``` perl
use configdata qw(%config);
```

dot-asm replied:
> I assume that this would also do better in add-depends.pl.

Yes. Under assumption that you actually venture into file canonization in add-depends.pl. Because I don't feel that one actually has to. And if one doesn't (or chooses to do it later), then suggested replacement would do some good. At least it would be a bit more consistent with Unix, which is customarily free from absolute paths.

dot-asm replied:
> use configdata qw(%config);

I suppose. I'm not that much into Perl internals, but if this doesn't instantiate a big fat hash, then yes, it would be totally appropriate.

levitte replied:
> Under assumption that you actually venture into file canonization in add-depends.pl. Because I don't feel that one actually has to.

*If* we're doing that, I would rather have it done in *one* place.

> (or chooses to do it later)

"later", is that "post 1.1.1"?  I mean, fixing such things in `add-depend.pl` can certainly be done before the final release...

> > ``` perl
> > use configdata qw(%config);
> 
> I suppose. I'm not that much into Perl internals, but if this doesn't instantiate a big fat hash, then yes, it would be totally appropriate.

Nah, I realised that makes no real difference...  It will still instanciate everything from that file, the LIST following the module name is just about what's getting imported into the current module (main::).  It's possible that perl will choose to free the unused stuff when it sees that nothing refers to, but...

dot-asm replied:
> If we're doing that, I would rather have it done in one place.

Sure, and one should be able to revert this replacement when/if you collect yourself to actually doing it [in that one magic place].

levitte replied:
Side note: Actually, it just dawned on me that we can as well do all this post-processing stuff in `add-depends.pl`

dot-asm replied:
Yeah. Nicest things would be that you'd do only one canonization per header file [and then reuse the result in each subsequent occurrence.]

levitte replied:
Ah, a result cache. That can be done.

dot-asm replied:
So what's final conclusion? Do you seek approval as it is, or do you choose to keep the suggested replacement till better times?

levitte replied:
For the moment, I'm away from the computer... but I'm pondering moving this all to `add-depends.pl` and work from there.

dot-asm replied:
I rather meant what's final conclusion for *this* request. It feels that add-depends.pl thing is matter for separate request...

levitte replied:
For this request, I'm done

levitte commented:
BTW, appveyor error is unrelated

richsalz commented:
I wonder if it would be easier to write a perl script that scanned our sources, and looked for OUR header files.

levitte commented:
You'd have to reimplement `makedepend`, and I really think that's taking it a bit far...

richsalz commented:
Yeah, that was my point.  It's actually pretty easy to do the job good enough for us.  But I lost my 50 line script, so oh well.

dot-asm commented:
I've removed 1.1.0 label, because there is no util/add-depends.pl.

levitte commented:
Squashed.  Merged!  Thanks you.

23be743286c0f0a160de33365ef34af39427eac9 util/postprocess-makedepend.pl: make an effort to collect dependencies
27c4490521261f7f53f494140d8a2b2a239306ef util/postprocess-makedepend.pl: For VC, don't include system headers
bbb9413b7e530aa25779b33b5ca7a5fd610e66f1 Streamline dependency generation
